### PR TITLE
Catalyst model tracing preview

### LINKED ISSUE #180: [feature] add jit.trace support
Add automated model tracing & packing.

### PR DESCRIPTION
**Pull request by hexfaker**
Fixes #180

### CODE REVIEW

File: catalyst/dl/scripts/trace.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    Model that calls specified method instead of forward
+
+    (Workaround, sigle method tracing is not supported)
+    """
+    def __init__(self, model, method_name):
+        super().__init__()
+        self.model = model
+        self.method = method_name
+
+    def forward(self, *input):
```
Comment (Scitator) [inline catalyst/dl/scripts/trace.py]: what about `*args, **kwargs` usage?

File: catalyst/dl/experiments/experiment.py [hunk 2]
```diff
@@ -70,13 +69,18 @@ def get_model_stuff(self, model, stage: str):
     def get_callbacks(self, stage: str) -> "List[Callback]":
         pass
 
+    @abstractmethod
```
Comment (Scitator) [inline catalyst/dl/experiments/experiment.py]: all we need during training is loaders, so I am not sure it is good idea to force user to implement `get_datasets` and `get_datasets_only`

File: catalyst/dl/scripts/trace.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    model.load_state_dict(torch.load(checkpoint_path)["model_state_dict"])
+    model.eval()
+
+    for p in model.parameters():
+        p.requires_grad_(False)
+
+    tracer = _TracingModelWrapper(model, method_name)
+    runner: Runner = runner_type(tracer.cpu(), torch.device("cpu"))
+
+    train_dataset = experiment.get_datasets_only(stage)["train"]
```
Comment (Scitator) [inline catalyst/dl/scripts/trace.py]: we do not know datasets names, so I suggest use something like
```
datasets.key()[0]
```

Comment (Scitator) [reply catalyst/dl/scripts/trace.py]: and what about get loaders, take first one and get dataset from it?

**Review by Scitator — CHANGES_REQUESTED** [review]

File: catalyst/dl/experiments/experiment.py (hunk 2 above)
Comment (hexfaker) [reply catalyst/dl/experiments/experiment.py]: `get_datasets_only` is implemented in `ConfigExperiment` and users should not reimplement it in most cases.  
May be we should update our config structure to separrate num workers, batch size etc from the rest of data params to make get_loaders and get_datasets independent?

File: catalyst/dl/scripts/trace.py (hunk 3 above)
Comment (hexfaker) [reply catalyst/dl/scripts/trace.py]: Getting loaders was in the first version. But it was tremendously slow, due to large batch and slow worker initialization. Another option is to override those params, but Im not sure whether  it's a better option.

Comment (Scitator) [reply catalyst/dl/scripts/trace.py]: Even it is slow, I dont think it is an issue - you need to trace you model only once per several training usually. But some additional logic with `get_datasets_only` will force you to think about API each experiment :) (like "why do we need this method?")

another issue, with this new method - we will need to maintain it in future versions and check that it works correctly. 

solution with loaders does'nt have such issues (btw, you always can dynamically change `num_workers` and `batch_size` for speedup), that is why it looks more general for me

Comment (hexfaker) [reply catalyst/dl/scripts/trace.py]: OK, this time you're right 😋

**Review by Scitator — APPROVED** [review]

**Pull request merged**